airflow/io/__init__.py

```diff
@@ -0,0 +1,270 @@
+        self._location = location
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """Return the total length of the file, in bytes."""
+
+    @property
+    def location(self) -> str:
+        """The fully-qualified location of the output file."""
+        return self._location
+
+    @abstractmethod
+    def exists(self) -> bool:
+        """Check whether the location exists.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+        """
+
+    @abstractmethod
+    def to_input_file(self) -> InputFile:
+        """Return an InputFile for the location of this output file."""
+
+    @abstractmethod
+    def create(self, overwrite: bool = False) -> OutputStream:
+        """Return an object that matches the OutputStream protocol.
+
+        Args:
+            overwrite (bool): If the file already exists at `self.location`
+                and `overwrite` is False a FileExistsError should be raised.
+
+        Returns:
+            OutputStream: An object that matches the OutputStream protocol.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+            FileExistsError: If the file at self.location already exists and `overwrite=False`.
+        """
+
+
+class FileIO(ABC):
+    """A base class for FileIO implementations."""
+
+    @abstractmethod
+    def new_input(self, location: str, conn_id: str | None) -> InputFile:
+        """Get an InputFile instance to read bytes from the file at the given location.
+
+        Args:
+            location (str): A URI or a path to a local file.
+            conn_id (str): The connection ID to use when getting the file.
+        """
+
+    @abstractmethod
+    def new_output(self, location: str, conn_id: str | None) -> OutputFile:
+        """Get an OutputFile instance to write bytes to the file at the given location.
+
+        Args:
+            location (str): A URI or a path to a local file.
+            conn_id (str): The connection ID to use when getting the file.
+        """
+
+    @abstractmethod
+    def delete(self, location: str | InputFile | OutputFile, conn_id: str | None) -> None:
+        """Delete the file at the given path.
+
+        Args:
+            location (Union[str, InputFile, OutputFile]): A URI or a path to a local file--if an InputFile
+                instance or an OutputFile instance is provided, the location attribute for that instance is
+                used as the URI to delete.
+            conn_id (str): The connection ID to use when deleting the file.
+
+        Raises:
+            PermissionError: If the file at location cannot be accessed due to a permission error.
+            FileNotFoundError: When the file at the provided location does not exist.
+        """
+
+
+FSSPEC_FILE_IO = "airflow.io.fsspec.FsspecFileIO"
+
+
+def _import_file_io(io_impl: str) -> FileIO | None:
```

Comment:
Where is this used? I tried to search but couldn’t find anything.

Reply:
Also this seems to be functionally the same as `import_string` with instantiation.

Reply:
Not used yet - it comes from the port. Yes, this needs to be adjusted to use the Airflow API for this.